docs/reference/indices/diskusage.asciidoc

```diff
@@ -0,0 +1,194 @@
+++++
+<titleabbrev>Analyze index disk usage</titleabbrev>
+++++
+
+experimental[]
+
+Analyzes the disk usage of each field of an index. This API
+might not support indices created in previous {es} versions.
+The result of a small index can be inaccurate as some parts
+of an index might not be analyzed by the API. It's recommended
+to execute this API with a single index (or the latest backing
+index of a data streams) as the API consumes resources significantly.
+
+[source,console]
+--------------------------------------------------
+POST /my-index-000001/_disk_usage?run_expensive_tasks=true
+--------------------------------------------------
+// TEST[setup:messages]
+
+[[analyze-index-disk-usage-api-request]]
+==== {api-request-title}
+
+`POST /<index>/_disk_usage`
+
+[[analyze-index-disk-usage-api-request-prereqs]]
+==== {api-prereq-title}
+
+* If the {es} {security-features} are enabled, you must have the `manage`
+<<privileges-list-indices,index privilege>> for the target index or index alias.
+
+[[analyze-index-disk-usage-api-path-params]]
+==== {api-path-parms-title}
+
+include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=index]
+
+[[analyze-index-disk-usage-api-query-params]]
+==== {api-query-parms-title}
+
+include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=allow-no-indices]
++
+Defaults to `true`.
+
+include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=expand-wildcards]
++
+Defaults to `open`.
+
+include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=index-ignore-unavailable]
+
+include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=wait_for_active_shards]
+
+include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=timeoutparms]
+
+`run_expensive_tasks`::
+(Optional, Boolean)
+Since analyzing the per field disk usage of indices is expensive and resource-intensive,
+the parameter `run_expensive_tasks` must be set to `true` in order for the task to be
+performed.
+Defaults to `false`.
+
+`flush`::
+(Optional, Boolean)
+If `true`, the target index will be flushed before the analyzing process starts.
+If `false`, then uncommitted data might not be included in the result.
+Defaults to `true`.
+
+
+[[analyze-index-disk-usage-api-example]]
+==== {api-examples-title}
+
+The following request analyzes the disk usage of index `my-index-000001`.
+
+[source,console]
+--------------------------------------------------
+POST /my-index-000001/_disk_usage?run_expensive_tasks=true
+--------------------------------------------------
+// TEST[setup:messages]
+
+The API returns the following response:
+
+[source,console-response]
+--------------------------------------------------
+{
+    "_shards": {
+        "total": 1,
+        "successful": 1,
+        "failed": 0
+    },
+    "my-index-000001": {
+        "index_size": "929mb", <1>
+        "index_size_in_bytes": 974192723,
+        "all_fields": {
+            "total": "928.9mb", <2>
+            "total_in_bytes": 973977084,
+            "terms": "107.8mb",
```

Comment:
I reached out to Adrien to discuss this:
- In his proposal `inverted_index` would be a container for all the fields that belong to the inverted index structures, namely terms, postings, frequencies, positions, offsets, payloads. The field usage API would then show the information as:

```
"inverted_index" : {
    "terms" : 1,
    "frequencies" : 1,
    "positions" : 0,
    "offsets" : 0,
    "payloads" : 0
},
"doc_values" : 1,
"points" : 0,
...
```

while the disk usage API would group the information under the same  "inverted_index"  section, but with possibly differently named subfields (or even no subfield at all in a first step). In a first step, the disk usage API could just show a total here, not a more refined view (as that would require introducing more concepts/names for what's currently the combined terms and frequencies disk usage). This would already help users make a choice between "index: true | false" for a field. 
- Another option we discussed would be for field usage to carry the  "terms", "frequencies", "positions" etc. fields, and the disk usage would then use combined field names such as "terms+frequencies" to make it clear that the metric shown is displaying disk usage from both of these combined. The field names would look as follows:

```
field_usage:

- "terms"
- "frequencies"
- "positions"
- "offsets"
- "payloads"
- "stored_fields"
- "doc_values"
- "points"
- "norms"
- "term_vectors"

disk_usage:

- "terms+frequencies"
- "positions+offsets+payloads"
- "stored_fields"
- "doc_values"
- "points"
- "norms"
- "term_vectors"
```

- It's also possible to combine both proposals by putting these combined field names under an "inverted_index" section.

Finally we discussed that disk usage could have multiple modes of executing, a quicker and a more expensive one, where the more expensive one could provide a more detailed breakdown of frequencies vs terms storage for example by rewriting the index.

Reply:
@ywelsch @jpountz Thanks for discussing it. I like the name `inverted_index` most and in this iteration, it just shows the total (i.e., no subfields).

> Finally we discussed that disk usage could have multiple modes of executing, a quicker and a more expensive one, where the more expensive one could provide a more detailed breakdown of frequencies vs terms storage for example by rewriting the index.

+1. I will work on it in a follow-up.